### PR TITLE
Set - 15: Realm objects

### PR DESCRIPTION
**Pull request by clementetb**
Adds support for realm models.

Note: Realm object sets don't support null values.

### CODE REVIEW

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java [hunk 1]
```diff
@@ -364,7 +364,7 @@ public void compileRequiredTypes() throws IOException {
     public void compileInvalidRequiredTypes() throws IOException {
         final String[] invalidRequiredAnnotationFieldTypes = {"byte", "short", "int", "long", "float", "double",
                 "boolean", "RealmList<Simple>", "Simple", "Mixed", "RealmList<Mixed>", "RealmDictionary<Simple>",
-                "RealmDictionary<Mixed>"};
+                "RealmDictionary<Mixed>", "RealmSet<Simple>", "RealmSet<Mixed>"};
```
Comment (edualonso) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java:367]: Does this test pass? I think it should fail unless you add support for `RealmSet<Mixed>` in the annotation processor 🤔

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+        private val setGetter: KFunction1<AllTypes, RealmSet<T>>,
+        private val setSetter: KFunction2<AllTypes, RealmSet<T>, Unit>,
+        private val managedSetGetter: KProperty1<SetContainerClass, RealmSet<T>>,
+        private val managedCollectionGetter: KProperty1<SetContainerClass, RealmList<T>>,
+        private val unmanagedInitializedSet: List<T?>,
+        private val unmanagedNotPresentValue: T,
+        private val toArrayManaged: ToArrayManaged<T>
+) : SetTester {
+
+    private lateinit var managedTester: ManagedSetTester<T>
```
Comment (edualonso) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:46]: Interesting approach using delegation. We could in theory have just a few delegates depending on the data type we are testing instead of the "asserters" I used for dictionaries.

Just for sanity, write a comment about this, something like "we use the base tester class for those methods that require no further abstraction whereas for the rest we provide our own custom logic" or the like.

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java [hunk 3]
```diff
@@ -1069,6 +1077,116 @@ boolean retainAllInternal(Collection<?> c) {
...
+
+    public RealmModelSetOperator(BaseRealm baseRealm, OsSet osSet, Class<T> valueClass) {
+        super(baseRealm, osSet, valueClass);
+    }
+
+    @Override
+    boolean add(@Nullable T value) {
+        // RealmModel sets doesnt allow null values
+        if (value == null) {
+            throw new IllegalArgumentException("Null values cannot be added into this set.");
```
Comment (edualonso) [inline realm/realm-library/src/main/java/io/realm/SetValueOperator.java:1093]: This should throw a `NullPointerException` according to the Java `Set` interface: https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#add(E)

Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/SetValueOperator.java:1093]: You should in general double-check the `Set` interface for restrictions in parameters and how the should throw. I completely ignored these for dictionaries and had to change them last week 🙊

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java [hunk 4]
```diff
@@ -1069,6 +1077,116 @@ boolean retainAllInternal(Collection<?> c) {
...
+            RealmObjectProxy proxy = (RealmObjectProxy) ((copyObject) ? CollectionUtils.copyToRealm(baseRealm, (RealmModel) value) : value);
+            Row row$realm = proxy.realmGet$proxyState().getRow$realm();
+            return osSet.addRow(row$realm.getObjectKey());
+        }
+    }
+
+    @Override
+    boolean containsInternal(@Nullable Object value) {
+        if (value == null) {
+            return false; // Realm model sets cannot contain nulls
```
Comment (edualonso) [inline realm/realm-library/src/main/java/io/realm/SetValueOperator.java:1111]: Same as above, this should be `NullPointerException`.

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java [hunk 5]
```diff
@@ -1069,6 +1077,116 @@ boolean retainAllInternal(Collection<?> c) {
...
+            Row row$realm = ((RealmObjectProxy) value).realmGet$proxyState().getRow$realm();
+            return osSet.containsRow(row$realm.getObjectKey());
+        }
+        throw new IllegalArgumentException("Only managed models can be contained in this set.");
+    }
+
+    @Override
+    boolean removeInternal(@Nullable Object value) {
+        if (value == null) {
+            return false;  // Realm model sets cannot contain nulls
```
Comment (edualonso) [inline realm/realm-library/src/main/java/io/realm/SetValueOperator.java:1123]: Same as above.

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java [hunk 6]
```diff
@@ -1069,6 +1077,116 @@ boolean retainAllInternal(Collection<?> c) {
...
+    }
+
+    @Override
+    boolean containsAllInternal(Collection<?> collection) {
+        // Collection has been type-checked from caller
+        //noinspection unchecked
+        Collection<T> realmModelCollection = (Collection<T>) collection;
+
+        // Realm model sets cannot contain null values
+        if (collection.contains(null)) { return false; }
```
Comment (edualonso) [inline realm/realm-library/src/main/java/io/realm/SetValueOperator.java:1149]: Same as above.

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+
+    private fun initAndAssertEmptySet(realm: Realm = this.realm): RealmSet<T> {
+        val allTypesObject = createAllTypesManagedContainerAndAssert(realm, "unmanaged")
+        assertNotNull(allTypesObject)
+        val set = setGetter.call(allTypesObject)
+        assertTrue(set.isEmpty())
+        return set
+    }
+
+    override fun toString(): String = when (mixedType) {
```
Comment (edualonso) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:61]: This should not take into consideration `mixedType` since this tester should not be used for `Mixed` - unless you plan to do so...?

**Review by edualonso — CHANGES_REQUESTED** [review]
Looks terrific! There a few minor things but nothing serious. Let me know if you need a second pass when ready.

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+            assertFalse(set.addAll(sameValuesManagedSet as Collection<T>))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Does not change if we add an empty RealmSet
+            val emptyManagedSet = managedSetGetter.get(transactionRealm.createObject())
+            assertTrue(emptyManagedSet.isEmpty())
+            assertFalse(set.addAll(emptyManagedSet))
+            assertEquals(unmanagedInitializedSet.size, set.size)
+
+            // Changes after adding a managed set containing other values
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:279]: I think we would also need a test showing that adding the same unmanaged objects without primary keys twice actually adds multiple copies, as core uses the object id and wouldn't consider them the same.

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+        // Test with unmanaged realm objects
+        val set = initAndAssertEmptySet()
+        realm.executeTransaction {
+            // Check throws exception when unmanaged values are passed
+            assertFailsWith<IllegalArgumentException>("Collection with unmanaged objects not permitted") {
+                set.retainAll(unmanagedInitializedSet)
+            }
+
+            // Check throws exception when null values are passed
+            assertFailsWith<IllegalArgumentException>("Collections with nulls are not permitted") {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:341]: Following Eduardo's comment, I guess we should throw a NullPointerException to follow Java's Set API.

File: realm/realm-library/src/androidTest/kotlin/io/realm/entities/SetContainerClass.kt [hunk 10]
```diff
@@ -26,7 +26,7 @@ import java.util.*
 
 open class SetContainerClass : RealmObject() {
 
-//    val myRealmModelSet = RealmSet<StringOnly>()
+    val myRealmModelSet = RealmSet<DogPrimaryKey>()
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/entities/SetContainerClass.kt:29]: Maybe worth also to have some tests for objects without primary key, just to show that we are not using `equalsTo` and catch any accidentals regressions on that matter.

File: realm/realm-library/src/main/java/io/realm/CollectionUtils.java [hunk 11]
```diff
@@ -148,6 +171,31 @@ static Mixed copyToRealmIfNeeded(BaseRealm realm, Mixed mixed) {
...
+    /**
+     * Called by both list, sets and dictionary operators to copy a RealmModel to Realm in case it has
+     * been deemed necessary.
+     *
+     * @param baseRealm The Realm instance to copy the object to.
+     * @param objects   collection of object to copy.
+     * @param <E>       The RealmModel type.
+     * @return the copied object
+     */
+    public static <E extends RealmModel> Collection<E> copyToRealm(BaseRealm baseRealm, Collection<E> objects) {
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/CollectionUtils.java:183]: Maybe be explicit about it returning a list, as people could think that it was in fact the same type of collection as the argument.

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+        val set = initAndAssertEmptySet()
+        realm.executeTransaction {
+            // Adding a value for the first time returns true
+            unmanagedInitializedSet.forEach { value ->
+                assertTrue(set.add(value))
+            }
+            // Adding an existing value returns false
+            unmanagedInitializedSet.forEach { value ->
+                assertFalse(set.add(value))
+            }
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:155]: Is it caught in the preprocessor that we do not support embedded objects?

**Review by rorbech — COMMENTED** [review]
Know it is still draft, but added some minor comments here and there. As Eduardo already pointed out ensure to honor the set interface on invalid types.

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java (hunk 1 above)
Comment (clementetb) [reply realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java:367]: Yes, it does

File: realm/realm-library/src/main/java/io/realm/internal/objectstore/OsObjectBuilder.java [hunk 13]
```diff
@@ -835,6 +836,24 @@ public void addUUIDSet(long columnKey, RealmSet<UUID> set) {
         addSetItem(builderPtr, columnKey, set, uuidSetItemCallback);
     }
 
+    public <T extends RealmModel> void addObjectSet(long columnKey, @Nullable RealmSet<T> set) {
```
[the remark this answers is not in the document]
Comment (clementetb) [reply realm/realm-library/src/main/java/io/realm/internal/objectstore/OsObjectBuilder.java:839]: Yes indeed, I didn't realize it was there. Actually, all set object builder setters have the same logic as lists.

File: realm/realm-library/src/main/java/io/realm/CollectionUtils.java (hunk 11 above)
Comment (clementetb) [reply realm/realm-library/src/main/java/io/realm/CollectionUtils.java:183]: Actually it is not needed forgot to remove it

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt (hunk 12 above)
Comment (clementetb) [reply realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:147]: It would throw in runtime, when validating the schema. I have added a check in the processor to catch it.

File: realm/realm-library/src/androidTest/kotlin/io/realm/entities/SetContainerClass.kt (hunk 10 above)
Comment (clementetb) [reply realm/realm-library/src/androidTest/kotlin/io/realm/entities/SetContainerClass.kt:30]: I've added a test to cover it.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProcessor.kt [hunk 14]
```diff
@@ -338,4 +344,29 @@ class RealmProcessor : AbstractProcessor() {
 
         return allValid
     }
+
+    // Because library classes are processed separately, there is no guarantee that this method can
+    // see all of the classes necessary to completely validate all of the realm model sets.  If it can find
+    // the fully-qualified class, though, and prove that the generic type is an embedded object,
```
Comment (rorbech) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProcessor.kt:350]: Could we maybe reuse https://github.com/realm/realm-java/pull/7220/files#diff-314913e78729941373d8ebb61f525077f7f5cc070da10ea6776bf4dc30c297e9R2416 to derive whether external classes are embedded to always raise this at compile time?

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java [hunk 15]
```diff
@@ -378,6 +379,22 @@ public void compileInvalidRequiredTypes() throws IOException {
         }
     }
 
+    // Not supported RealmSets of embedded objects
+    @Test
+    public void compileEmbeddedObjectSet() throws IOException {
```
Comment (rorbech) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java:384]:
```suggestion
    @Test
    public void compileSetWithEmbeddedObjectNotSupported() throws IOException {
```

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt [hunk 16]
```diff
@@ -716,35 +716,41 @@ class RealmProxyClassGenerator(private val processingEnvironment: ProcessingEnvi
...
-            if (forMixed) {
-                emitStatement("OsSet osSet = proxyState.getRow\$realm().getMixedSet(%s)", fieldColKeyVariableReference(field))
-            } else if (forRealmModel) {
-                emitStatement("OsSet osSet = proxyState.getRow\$realm().getModelSet(%s)", fieldColKeyVariableReference(field))
-            } else {
-                emitStatement("OsSet osSet = proxyState.getRow\$realm().getValueSet(%s, RealmFieldType.%s)", fieldColKeyVariableReference(field), Utils.getValueSetFieldType(field).name)
+            when {
+                forMixed -> {
+                    // Add Mixed logic
+                    throw UnsupportedOperationException("Missing Mixed implementation")
```
Comment (rorbech) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt:722]: Is this because it is not targeted for this PR or just not done yet because it is awaiting merge of another PR? If not part of this PR then link to the issue that should address this.

File: realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedSetTests.kt [hunk 17]
```diff
@@ -34,7 +34,7 @@ class ParameterizedSetTests(
         fun testType(): List<SetTester> {
             return SetMode.values().map { type ->
                 when (type) {
-                    SetMode.UNMANAGED -> unmanagedSetFactory()
+                    SetMode.UNMANAGED -> arrayListOf()
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedSetTests.kt:37]: Just for local testing?!

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+                    assertFailsWith<IllegalArgumentException>("Cannot pass values from another Realm") {
+                        setSetter(alternativeObject, alternativeSet)
+                    }
+                }
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    private fun objectsFromOtherThread(block: (values: List<T?>) -> Unit) {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:363]: Maybe rename so that it is obvious that you are actually doing the tests and playing around with the looper. It is not clear at the call site that it is doing a heavy lift.

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+            }
+
+            // Check throws exception when null values are passed
+            assertFailsWith<java.lang.NullPointerException>("Collections with nulls are not permitted") {
+                set.removeAll(listOf(null))
+            }
+        }
+
+        // Test with objects from another realm
+        objectsFromOtherThread {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:334]: 🎉

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+
+    override fun clear() = managedTester.clear()
+
+    override fun freeze() = managedTester.freeze()
+
+    override fun setters() {
+        managedTester.setters()
+
+        // Test with objects from another realm
+        looperThread.runBlocking {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:347]: Can you also use `objectFromOtherThread` here?

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+                    assertFailsWith<IllegalArgumentException>("Cannot pass an object from another Realm instance") {
+                        block(values)
+                    }
+                }
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    private fun objectFromOtherThread(block: (values: T) -> Unit) {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:379]: Do you really need to variants of this?

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java [hunk 22]
```diff
@@ -1069,6 +1077,132 @@ boolean retainAllInternal(Collection<?> c) {
...
+        checkValidCollection(realmModelCollection);
+        NativeMixedCollection collection = NativeMixedCollection.newRealmModelCollection(realmModelCollection);
+        return osSet.collectionFunnel(collection, OsSet.ExternalCollectionOperation.REMOVE_ALL);
+    }
+
+    @Override
+    boolean retainAllInternal(Collection<?> c) {
+        // Collection has been type-checked from caller
+        // Realm model sets cannot contain null values
+        if (c.contains(null)) {
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/SetValueOperator.java:1195]: Maybe this check could go into `checkValidCollection`?

File: realm/realm-library/src/main/java/io/realm/CollectionUtils.java [hunk 23]
```diff
@@ -26,22 +26,43 @@
...
     }
 
+    /**
+     * Called by both list and set operators to determine whether a RealmModel is an embedded object
+     *
+     * @param realm  the Realm instance to check against.
+     * @param object the object to check.
+     * @return true if the object can be copied, false otherwise
+     */
+    static boolean isEmbedded(BaseRealm realm, RealmModel object) {
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/CollectionUtils.java:46]: If we knew that compilation would fail for sets of embedded objects (see other comment) I guess we could eliminate this and all the runtime checks.

**Review by rorbech — COMMENTED** [review]
Nice. Still a bit unclear about the missing parts related to Mixed. Will they be part of this PR or will it be addresses in another issue?

There is also a suspicious change in https://github.com/realm/realm-java/pull/7403/files#diff-dabaa4d9a9a75c42771c46f9d263feb5bf1e1cfeb10469507d2644c8d03f5b8bR37 where the initialization is changed. Is this intentional because you now also test the setters or just left overs from local testing. 

I also didn't know you would do a full tester for the non-primary key case. Thought we chatted about just going with very explicit documentation and I think you could have gone away with some minor tests as part of the `add`/`remove` tests to highlight and catch eventual regressions. Sorry for not being more explicit - but I guess now we have the full-blown tests 🎉

**Comment by clementetb** [comment]
Mixed will be addressed in another PR.

File: realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedSetTests.kt (hunk 17 above)
Comment (clementetb) [reply realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedSetTests.kt:37]: Indeed, forgot to reenable

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt (hunk 16 above)
Comment (clementetb) [reply realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt:722]: Mixed is not part of this PR, it will be addressed on a separate one.

File: realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt (hunk 21 above)
Comment (clementetb) [reply realm/realm-library/src/androidTest/kotlin/io/realm/RealmModelManagedSetTester.kt:379]: Thats true, I have tidied up the code

**Review by rorbech — APPROVED** [review]
